Add DBR_VFIELD

mdavidsaver opened the pull request:
While not acceptable in it present form, I would like to get feedback before another iteration.  I really like to see a solution merged.

The problem I'm trying to solve is how to usefully access complex types as database fields.  My immediate application is to the PVA data containers `shared_vector` and `PVStructure`, although there are others (eg. areaDetector's NDArray).

I'd like to be able to move these types through `dbGet()` and `dbPut()`, and by extension the derivative interfaces like links.

I want to do this is a way which would avoid allocating a new `DBF_*` code for every possibility as a way to avoid close coupling of these (currently external) types into the database code.  My solution to this is to add one new `DBR_*` code, and use it to pass extra arguments around most of the existing code of `dbGet()` and `dbPut()`.  The logic to handle this is split between the caller, and record support.

So a caller would look like:

```c++
DBADDR *paddress = ...;
shared_vector<const void> aval;
struct VField {
    const VFieldType *vtype;
    shared_vector<const void> *aval;
} varg = {&vfSharedVector, &aval};
dbGet(paddress, DBR_VFIELD, NULL, &varg, NULL);
```

So in effect, `DBR_VFIELD` indicates that additional arguments are passed through the existing 'pbuffer' argument.  At present this ugliness is left exposed, though should eventually be hidden in a wrapper.

The other side of `dbGet()` is one of two new record support functions

```c++
long get_vfield(struct dbAddr *paddr, struct VField *p)
{
    stdstringinRecord *prec = (stdstringinRecord*)paddr->precord;

    if(p->vtype==&vfStdString) {
        VString *pstr = (VString*)p;
        if(dbGetFieldIndex(paddr)==stdstringinRecordVAL) {
            pstr->value = prec->val;
            return 0;
        } else if(dbGetFieldIndex(paddr)==stdstringinRecordOVAL) {
            pstr->value = prec->oval;
            return 0;
        }
    }
    return S_db_badChoice;
}
```

mdavidsaver committed:
add alarm message field

mdavidsaver committed:
use recGblSetSevrMsg()

mdavidsaver committed:
dbCommon add UTAG

mdavidsaver committed:
pass amsg/utag through dbGet() options

mdavidsaver committed:
add utag filter

mdavidsaver committed:
propagate utag w/ links (including TSEL)

add lset::getTimestampTag()

mdavidsaver committed:
add db_field_log::mask

Allow filters to distingush DBE_PROPERTY

mdavidsaver committed:
VField

mdavidsaver commented on modules/database/src/ioc/dbStatic/recSup.h:
```diff
@@ -52,6 +53,8 @@ struct typed_rset {
     long (*get_graphic_double)(struct dbAddr *paddr, struct dbr_grDouble *p);
     long (*get_control_double)(struct dbAddr *paddr, struct dbr_ctrlDouble *p);
     long (*get_alarm_double)(struct dbAddr *paddr, struct dbr_alDouble *p);
+    long (*get_vfield)(struct dbAddr *paddr, struct VField *p);
```
How to incorporate `db_field_log`?

anjohnson commented on modules/database/src/ioc/db/dbAccess.c:
```diff
@@ -625,6 +625,8 @@ long dbEntryToAddr(const DBENTRY *pdbentry, DBADDR *paddr)
     paddr->field_size  = pflddes->size;
     paddr->special     = pflddes->special;
     paddr->dbr_field_type = mapDBFToDBR[dbfType];
+    paddr->ro = 0;
```
You implied that `ro` doesn't mean the data is read-only, just that dbAccess can't write to such fields itself. Will there be fields that pvAccess can read directly but not write to, or do you expect all accesses to these to go through the put/get_vfield() methods?

Maybe `vo` for Virtual Only might be appropriate in that case, but having a non-NULL `vfields` pointer would probably be equivalent in that case, so your answer to the above question may be yes.

mdavidsaver replied:
The distinction is finer than "Virtual Only".  This can be accomplished with `DBF_NOACCESS`.  The `ro` flag disallows `dbPut()` when `dbr!=DBR_VFIELD`.  `dbGet()` is still allowed.

> ... do you expect all accesses to these to go through the put/get_vfield() methods?

Excepting device support, yes.  I see the present habit of going around `dbPut()`/`dbGet()` as a symptom of an interface which has grown inadequate.

anjohnson commented on modules/database/src/std/rec/stdstringinRecord.cpp:
```diff
@@ -0,0 +1,240 @@
+
+    // we don't allow dbPut() via DBF_CHAR.
+    // std::string::c_str() warns of dire consequences following modifications
+    paddr->ro = 1;
```
I think you can have dbAccess prevent puts by setting `paddr->special = SPC_NOMOD` here, but I don't know what that would do with your `put_vfield()` additions.

mdavidsaver replied:
Would this actually prevent `dbPut()` from working?  From my reading of the code, I don't think so.  At least no without  Also, I still want to distinguish fields which `dbPut()` should never be allowed (even through `put_vfield()`).

anjohnson commented on modules/database/src/std/dev/devSSiSoft.cpp:
```diff
@@ -0,0 +1,91 @@
+
+#ifndef USE_TYPED_DSET
+#  define USE_TYPED_DSET
+#endif
```
You don't need to do this at all, the modules/database/src/std/Makefile sets both RSET and DSET flags for all Base device support and record types.

anjohnson commented on modules/database/src/ioc/dbStatic/dbFldTypes.h:
```diff
@@ -87,6 +101,7 @@ mapdbfType pamapdbfType[DBF_NTYPES] = {
 #define DBR_ENUM        DBF_ENUM
 #define DBR_PUT_ACKT    DBR_ENUM+1
 #define DBR_PUT_ACKS    DBR_PUT_ACKT+1
+#define DBR_VFIELD      (DBR_PUT_ACKS+1)
```
How about `DBR_QFIELD`, `QField`, `get_qfield()`, `put_qfield()` etc.? You already own the letter Q in the EPICS world (and it has a bigger Scrabble score)…

mdavidsaver replied:
Wow.  1/26th of the alphabet.  Such an honor!

anjohnson reviewed (commented):
Some suggestions inline…

mdavidsaver committed:
AMSG feature test

mdavidsaver committed:
add lset::getAlarmMsg()

mdavidsaver committed:
add dbDbLinkTest

mdavidsaver committed:
UTAG uint64

mdavidsaver committed:
Split out amsg/utag options through dbGet()

KathrynBaker committed:
Addying Doxugen annotations for cvtFast.h and freeList.h

mdavidsaver committed:
dbfl_type_rec goes away

marciodo committed:
iocsh: set PWD

Set the env variable PWD during libCom IOC shell function registration.
Change environment variable PWD when a user calls cd from the IOC shell

mdavidsaver committed:
review cleanup

mdavidsaver committed:
doc

mdavidsaver committed:
minor and doc

mdavidsaver commented:
Superseded by https://github.com/epics-base/epics-base/pull/154
